### PR TITLE
Strapi promotions: Hebrew-independent localization, sidebar-ad fixes, and a deterministic Playwright suite

### PR DESCRIPTION
**Pull request by relyks**
> **Stacked series:** #3621 (the selection fix) is based on this branch and should be reviewed after it. This branch deliberately includes three `test.fail` markers documenting a selection bug that it does not fix. #3621 fixes the bug and converts those markers into passing tests. See "What could warrant closer review."

### Description

Sefaria shows promotional content managed in Strapi through banners, modals, and sidebar ads. This PR makes two main changes to that system.

1. **Fixes three production bugs found while adding tests**
   
   - **Sidebar ads missing on first page load.** The ad-building effect depended on a flag indicating that data had arrived instead of the data itself. Because React 16 does not batch state updates made inside promise callbacks, the effect could run while the data was still `null` and never run again. It now depends directly on the data it reads.
   - **Country targeting always using the English locale's settings.** `countriesToTarget` can differ by locale in Strapi, but the merge previously treated it as a shared field and kept the first row's value, which was always English. It is now included in `LOCALIZED_FIELDS`, and the display logic reads the value for the active locale.
   - **Hebrew and English sidebar ads interfering with each other.** Sidebar ads are now kept as separate locale-specific entries instead of being merged.

2. **Adds deterministic Playwright coverage for the Strapi integration.**
   
   The suite contains tests that do not depend on live Strapi data.
   
   Fourteen recorded HTTP archives replay real responses captured from a local Strapi instance. These verify that the app handles payloads Strapi actually produces.
   
   A synthetic payload factory covers cases that are difficult or impossible to represent through recordings, including multi-document permutations, unusual field values, error responses, and v4-shaped payloads.
   
   A contract spec compares the synthetic factory's fields against the committed recordings so the synthetic payload shape cannot silently drift away from actual Strapi responses.

### What could warrant closer review

1. **The three **`test.fail`**** markers are deliberate.** They assert the desired behavior for a bug this branch documents but does not fix. #3621, which is stacked on this branch, fixes the selection logic and converts them into normal passing tests.

2. **`LOCALIZED_FIELDS`** is important and easy to get wrong later.** Any Strapi field that can vary by locale needs to be added to it. Otherwise, the first row's value silently wins and the other locale's value is lost. This was the cause of the country-targeting bug.

3. **The HAR recordings have a maintenance cost.** They match both the full URL and POST body, so changing the GraphQL query in `context.js` invalidates all fourteen recordings. The synthetic factory exists partly to avoid relying on recordings for every permutation, and `tests/README.md` documents the re-recording workflow.

4. **The sidebar-ad first-load bug depends on React 16 behavior.** React 16 does not automatically batch state updates inside promise callbacks. A future React 18 upgrade would likely hide the original symptom, but the effect dependency fix is still correct. The first-load test should continue protecting the intended behavior through that upgrade.

### Key Design Decisions

- **Treat ****`LOCALIZED_FIELDS`**** as part of the merge contract.**
  
  Fields that editors can vary by locale are retained separately for each locale. Everything else is assumed to be identical and is taken from the first row.
  
  This means any field that can vary by locale must be included in `LOCALIZED_FIELDS`. Otherwise, the value from the first row silently replaces the value from the other locale. The code now documents this constraint.

- **Record real payloads before abstracting them.**
  
  Fourteen representative Strapi states were recorded before the synthetic payload factory was created. The factory's shape is therefore based on payloads Strapi actually produced rather than an assumed schema.
  
  The contract spec then compares the factory against those recordings. At least one recording for each content type must remain committed or the coverage guard fails.

- **Use recordings and synthetic payloads for different purposes.**
  
  HAR recordings verify behavior against real Strapi responses, but they match the full URL and POST body. A query change therefore invalidates them.
  
  Synthetic routes match the endpoint instead and are better suited for combinations that would be expensive or impossible to record, including multi-document permutations, unusual field values, precise timing cases, HTTP errors, non-JSON responses, and v4-shaped wrappers.

- **Keep this suite separate from the normal e2e page helpers.**
  
  The standard `goToPageWithLang` and `goToPageWithUser` helpers install overlay suppression that intercepts the Strapi endpoint with an empty payload and marks banners and modals as already seen. That removes the content this suite needs to test.
  
  These specs therefore use bare `page.goto` calls with their own routes. `CLAUDE.md` rule 22 documents this exception so the tests are not later converted to the normal PageManager pattern and unintentionally stop testing Strapi content.

- **Control the four inputs that make these tests deterministic.**
  
  The details are documented in `tests/README.md`, but the important constraints are:
  
  - *Country* comes from the `cf-ipcountry` header, not the browser timezone. Changing `timezoneId` would also change date query parameters and invalidate the HAR recordings.
  - *Clock* uses `clock.install()` together with `pauseAt()`. `install()` alone allows time to continue advancing, which makes `showDelay` boundary tests unreliable.
  - *Locale* comes from the `interfaceLang` cookie on the browser context. The `/interface/<lang>` route is avoided because it can redirect off-domain.
  - *Page path* must not match a recording's `buttonURL` pathname, because the excluded-path logic would suppress the promotion and make an absence assertion pass for the wrong reason.
  - Boundary tests also use `waitForTimerArmed`. The `showDelay` timer is created several async steps after the Strapi response arrives, so advancing the clock before the timer exists can make a "still hidden" assertion pass without actually testing the timer.

- **Represent the known selection bug with expected failures.**
  
  Three `test.fail()` cases in `strapi-selection-order.spec.js` document a bug this branch does not fix. Selection currently chooses a banner or modal based on the date window before applying locale and country targeting. This means the selected item can later be rejected even when another eligible item was available.
  
  The tests assert the desired behavior and are paired with passing controls showing that the individual content renders correctly. #3621 fixes the selection logic and converts these expected failures into normal passing tests.

### Code Changes

| File                                                            | Purpose                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| --------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `static/js/Promotions.jsx`                                      | Fixes first-load sidebar ads by making the effect depend on the data it reads.                                                                                                                                                                                                                                                                                                                                                                 |
| `static/js/Misc.jsx`                                            | Updates banner/modal display logic for the merged document shape and reads `countriesToTarget` from the active locale.                                                                                                                                                                                                                                                                                                                         |
| `static/js/sefaria/tests/strapiLocalization.test.js`            | Unit coverage for locale merging, localized fields, country targeting, and sidebar-ad fan-out.                                                                                                                                                                                                                                                                                                                                                 |
| `playwright.config.ts`                                          | Adds the `chrome-strapi` project for the Strapi-specific e2e suite and uses `SANDBOX_URL` directly as the base URL.                                                                                                                                                                                                                                                                                                                            |
| `e2e-tests/support/strapi-har-fixture.js` **(new)**             | Wraps `routeFromHAR` and fails when a request does not match the recording instead of falling through to the network.: `replayHarForSandboxOrigin` rewrites each recorded entry's origin to match `SANDBOX_URL` at replay time, so CI (a different origin than the recordings) can actually reach a matching request; and the afterEach guard now also fails when nothing was served at all, not only when something was served but unmatched. |
| `e2e-tests/support/strapi-payload-factory.js` **(new)**         | Generates synthetic Strapi payloads using field defaults based on the recordings, including document IDs, locale rows, country targeting, and date helpers.                                                                                                                                                                                                                                                                                    |
| `e2e-tests/support/strapi-payload-fixture.js` **(new)**         | Provides `routeWithStrapiPayload` and `expectStrapiServed`, including a guard that fails if the page never requested the synthetic payload.                                                                                                                                                                                                                                                                                                    |
| `e2e-tests/tests/strapi.fixtures.js` **(new)**                  | Contains the recorded scenarios and shared test helpers, including `prepareStrapiPage`, clock helpers, `waitForTimerArmed`, and `useInterfaceLanguage`.                                                                                                                                                                                                                                                                                        |
| `e2e-tests/tests/strapi-*.spec.js` + `e2e-tests/fixtures/*.har` | The 19 Playwright spec files and 14 recorded Strapi responses.                                                                                                                                                                                                                                                                                                                                                                                 |
| `e2e-tests/tests/README.md`, `e2e-tests/CLAUDE.md`              | Documents the recording workflow, test controls, coverage, recorded vs. synthetic guidance, and why the suite does not use the normal PageManager helpers.                                                                                                                                                                                                                                                                                     |

### Tests

<details>
<summary><strong>Recorded scenarios: 34 tests across 14 recordings</strong></summary>

Each recording represents a real content state captured from a local Strapi instance and is pinned to the clock under which it was recorded. The specs reference scenarios rather than hardcoding their content.

- **Locale separation across banners, modals, and sidebar ads** (`strapi-banner.spec.js`, `-banner-hebrew`, `-banner-bilingual`, `-modal`, `-modal-hebrew`, `-modal-bilingual`, `-sidebar-ad-hebrew`, `-sidebar-ad-bilingual`)
  
  English-only content renders under the English interface but not Hebrew, and Hebrew-only content does the reverse. Bilingual content shows each interface only its own localized copy and button URL.
  
  The bilingual modal also tests an optional-header difference where `modalHeader` exists in Hebrew but is `null` in English. The Hebrew version renders an `<h1>`, while the English version does not create one.

- **Country targeting by locale** (`-banner-country-targeting`, `-modal-country-targeting`)
  
  Covers include and exclude targeting through the `cf-ipcountry` header, including cases where the two locales have different targeting rules.

- **Date windows** (`-banner-expired`, `-banner-future`, `-sidebar-ad-date-states`)
  
  Expired and future banners are delivered but not rendered. Sidebar ads cover active, expired, and future states together because they apply date filtering independently inside `Promotions`.

- **Dismissal persistence** (`-banner`)
  
  Dismissing a banner keeps it hidden after reload. Clearing the persisted dismissal key makes it appear again.

- **Keyword targeting** (`-sidebar-ad`)
  
  Sidebar ads appear on matching topic-category pages and are suppressed on excluded pages using the real `keywords` to page-context matching.

</details>

<details>
<summary><strong>Synthetic specs: 38 tests</strong></summary>

- **`strapi-payload-contract.spec.js`**
  
  Compares the factory's fields, nesting, and types against every committed recording for each content type. It also requires at least one recording for every content type.

- **`strapi-show-delay.spec.js`**
  
  Verifies the `showDelay` boundary for each surface, including one second before, one second after, and multiple surfaces with different delays in the same payload.
  
  These tests use `waitForTimerArmed` so the clock is not advanced before the timer exists.

- **`strapi-selection-order.spec.js`**
  
  Tests which document wins when multiple documents are within their date windows. Identical documents fall back to payload order, with a reversal test confirming that behavior.
  
  It also contains the three expected-failure cases for the known selection bug. Passing controls show that the same content renders when tested independently.

- **`strapi-excluded-paths.spec.js`**
  
  Verifies that a surface is suppressed on a page matching its own button URL, including cases where the other locale's URL matches. Each suppression test has a control that changes only the page URL.

- **`strapi-payload-resilience.spec.js`**
  
  Covers empty responses, v4-wrapper-shaped responses, HTTP 500s, and non-JSON responses. Each case verifies that the page remains usable without promotions and that failures are reported once.

</details>

<details>
<summary><strong>Unit tests</strong></summary>

- **`sidebarAds.test.js`**
  
  Verifies that each locale row becomes its own locale-tagged sidebar ad without cross-locale interference.

</details>

### Testing Patterns

The suite has a few protections intended to make negative assertions meaningful.

HAR replay blocks unmatched Strapi requests instead of silently falling through to the network, and now also fails if nothing was served from the HAR at all. Synthetic tests use `expectStrapiServed` to fail if the page never requested the supplied payload.

This matters because many of the assertions verify that something did *not* render. Without these guards, a broken fixture could cause those tests to pass without exercising the feature at all.

The same principle applies to controls. Tests that assert a promotion is absent generally include either a sibling test showing that the same content renders under valid conditions or another surface in the same payload that should always render.

The three expected-failure tests follow the same pattern. They assert the behavior we actually want rather than encoding the current bug as a passing test. When #3621 fixes the bug, those assertions remain valid and only the `test.fail` markers need to be removed.

### CODE REVIEW

**Review by stevekaplan123 — COMMENTED** [review]
Claude points out that `14 HAR recordings invalidate together whenever the GraphQL query changes. Re-recording requires a local Strapi.  If anyone edits the GraphQL query in context.js, all 14 recordings stop matching and must be re-recorded against a local Strapi.`.  Is this risk real? 
Another Claude comment:
```
### `LOCALIZED_FIELDS` — which other fields are localized in Strapi?

The `countriesToTarget` fix is right. My question is about the fields it doesn't cover,
since the failure mode has no symptom: anything not in `LOCALIZED_FIELDS` comes from
`rows[0]`, so the other locale's value is silently dropped rather than erroring. With
`SUPPORTED_LOCALES = ["en", "he"]` that means the English row supplies the shared
fields whenever the document has one, and the Hebrew row supplies them when it's
Hebrew-only — so the same field is sourced from different languages depending on the
document, which is part of why this is hard to spot by eye.

The "identical across locale rows" assumption is already false in this PR's own
fixtures — in `strapi-banner-both-locales.har`, the two locale rows of the same
document have different timestamps:

    en_banners  createdAt=2026-02-11T21:05:12Z  publishedAt=2026-08-05T09:06:02Z
    he_banners  createdAt=2026-08-05T08:29:15Z  publishedAt=2026-08-05T08:43:45Z

That's Strapi working as designed (each locale publishes separately), and harmless
because nothing reads those fields. But it means the assumption isn't self-evident.

**The ask:** whether a field is localized is a per-field toggle in Strapi's
Content-Type Builder, and `LOCALIZED_FIELDS` is a hand-typed copy of it with nothing
linking the two. Could you check the content model and confirm which fields have
localization enabled? I'd look at the five `showTo*` fields first — they answer "who
may see this," the same question `countriesToTarget` answers, in the same content type.
If the answer is "only the ones now listed," worth saying so in the comment so the next
person doesn't have to re-derive it.
```

I'm happy to approve it if you think neither of these concerns is valid @relyks

**Comment by relyks** [comment]
> Claude points out that `14 HAR recordings invalidate together whenever the GraphQL query changes. Re-recording requires a local Strapi. If anyone edits the GraphQL query in context.js, all 14 recordings stop matching and must be re-recorded against a local Strapi.`. Is this risk real? Another Claude comment:
> 
> ```
> ### `LOCALIZED_FIELDS` — which other fields are localized in Strapi?
> 
> The `countriesToTarget` fix is right. My question is about the fields it doesn't cover,
> since the failure mode has no symptom: anything not in `LOCALIZED_FIELDS` comes from
> `rows[0]`, so the other locale's value is silently dropped rather than erroring. With
> `SUPPORTED_LOCALES = ["en", "he"]` that means the English row supplies the shared
> fields whenever the document has one, and the Hebrew row supplies them when it's
> Hebrew-only — so the same field is sourced from different languages depending on the
> document, which is part of why this is hard to spot by eye.
> 
> The "identical across locale rows" assumption is already false in this PR's own
> fixtures — in `strapi-banner-both-locales.har`, the two locale rows of the same
> document have different timestamps:
> 
>     en_banners  createdAt=2026-02-11T21:05:12Z  publishedAt=2026-08-05T09:06:02Z
>     he_banners  createdAt=2026-08-05T08:29:15Z  publishedAt=2026-08-05T08:43:45Z
> 
> That's Strapi working as designed (each locale publishes separately), and harmless
> because nothing reads those fields. But it means the assumption isn't self-evident.
> 
> **The ask:** whether a field is localized is a per-field toggle in Strapi's
> Content-Type Builder, and `LOCALIZED_FIELDS` is a hand-typed copy of it with nothing
> linking the two. Could you check the content model and confirm which fields have
> localization enabled? I'd look at the five `showTo*` fields first — they answer "who
> may see this," the same question `countriesToTarget` answers, in the same content type.
> If the answer is "only the ones now listed," worth saying so in the comment so the next
> person doesn't have to re-derive it.
> ```
> 
> I'm happy to approve it if you think neither of these concerns is valid @relyks

Yeah, both concerns are valid, but I don't think either should block these PRs. Neither breaks the intended functionality, and both can be addressed in follow-up PRs.

It is true that the HAR recordings become invalid when the GraphQL query changes. I was already aware of that limitation. The tests in both PRs currently use either HAR recordings or the mocked payload factory. If I change the fields in the Strapi content types, which will likely happen in the future, I will also need to update the GraphQL query. At that point, I will either need to re-record the HAR files or update those tests to use the mocked payload factory instead.

To create the mocked payload factory, I manually recreated each recorded scenario in my local Strapi instance. I then had Claude derive a function from those HAR recordings that could generate equivalent synthetic responses. For the future, I either need to move the remaining recorded tests to that factory or find a way to control Strapi programmatically. Programmatically controlling Strapi is probably the better long-term approach, but it would still require running a Strapi instance locally or in CI. I couldn't figure that out in time to finish this work for the deployment.

The purpose of both the HAR recordings and the mocked payload factory is to avoid making external API requests during testing. They mock the boundary between the frontend and Strapi, allowing tests to run without a Strapi instance. This also makes them much faster and deterministic.

For the second concern, localization is configured using a separate toggle for each field in Strapi. I rechecked, and only the fields listed in the code can be localized. Claude's claim that "the identical across locale rows assumption is already false" is incorrect. Fields that aren't localized remain the same across locale entries. Each locale entry can still be published separately, but that doesn't make every field locale-dependent.

A better long-term approach would be to query Strapi to determine which fields are localizable. That would keep the Strapi configuration and frontend code in sync without needing manual code changes whenever the configuration changes. I will investigate that in the future.

File: e2e-tests/CLAUDE.md [hunk 1]
```diff
@@ -48,6 +48,7 @@ Tests run against the sandbox URL pointed at by `SANDBOX_URL` / `SANDBOX_URL_IL`
...
... surfaces under test (`#bannerMessage`, `#interruptingMessageBox`, `.sidebarPromo`) also have no roles or labels to target, so CSS is the only available anchor. These specs never contact live Strapi. They get their payload one of two ways — replaying a committed HAR recording (`routeWithStrapiHarFixture`), or building one in code (`routeWithStrapiPayload` + `support/strapi-payload-factory.js`) for states that cannot be published or recorded: several items at once, unpublishable field values, and error responses. Both kinds carry an `afterEach` guard (`expectStrapiServedFromHar` / `expectStrapiServed`) that turns "the payload never reached the page" into an explicit failure instead of a silently passing absence assertion — keep it when editing. Before touching any of this read [tests/README.md](tests/README.md), which covers when to record versus synthesize, the recording workflow, the faked clock (`install()` alone is **not** enough — it keeps running with real time; `pauseAt` is what freezes it), `waitForTimerArmed` for `showDelay` boundaries, the `cf-ipcountry` country lever, the stale-fixture guard, and the payload contract spec that holds the factory to the real recorded shape.
```
[the remark this answers is not in the document]
Comment (relyks) [reply e2e-tests/CLAUDE.md:51]: This is technically true, but I'll have to come back to it in the future. I have no way of addressing that now. Image assets in Strapi Cloud do not work the same way as they do locally.

File: e2e-tests/tests/strapi-targeting.spec.template.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+/**
+ * Playwright Tests: Strapi Targeting & Date States
+ *
+ * Two orthogonal gates, driven without any mocking of application code:
+ *   - Country targeting: getViewerCountryCandidates() derives the viewer's plausible countries
+ *     from Intl.DateTimeFormat().resolvedOptions().timeZone (countryCandidates.js). So Playwright's
+ *     `timezoneId` context option IS the country lever. A modal that INCLUDE-targets the US shows
+ *     for a US timezone and is hidden for a non-US timezone.
```
[the remark this answers is not in the document]
Comment (relyks) [reply e2e-tests/tests/strapi-targeting.spec.template.js:8]: This will be addressed in the future when I remove the HAR recordings or figure out a different method as I mentioned in my comment below.

[the remark this answers is not in the document]
Comment (yitzhakc) [reply e2e-tests/tests/strapi-targeting.spec.template.js:8]: Do we need to add this to a story somewhere?

**Comment by yitzhakc** [comment]
> Yeah, both concerns are valid, but I don't think either should block these PRs. Neither breaks the intended functionality, and both can be addressed in follow-up PRs.
Very good.
@relyks Please make sure that we have those written up as stories so we don't forget it
I'm approving in the meantime

**Review by yitzhakc — APPROVED** [review]

**Review by stevekaplan123 — APPROVED** [review]
